components/encryption/src/encrypted_file/mod.rs

```diff
@@ -0,0 +1,106 @@
+use std::fs::{rename, OpenOptions};
+use std::io::{ErrorKind, Read, Write};
+use std::path::Path;
+
+use kvproto::encryptionpb::EncryptedContent;
+use protobuf::Message;
+use rand::{thread_rng, RngCore};
+
+use crate::master_key::*;
+use crate::{Error, Result};
+
+mod header;
+use header::*;
+
+const TMP_FILE_SUFFIX: &str = ".tmp";
+
+/// An file encrypted by master key.
+pub struct EncryptedFile<'a> {
+    base: &'a Path,
+    name: &'a str,
+}
+
+impl<'a> EncryptedFile<'a> {
+    /// New an `EncryptedFile`.
+    ///
+    /// It's different from `std::fs::File`, it does not hold a reference
+    /// to the file or open the file, util we actually read or write.
+    pub fn new(base: &'a Path, name: &'a str) -> EncryptedFile<'a> {
+        EncryptedFile { base, name }
+    }
+
+    /// Read and decrypt the file.
+    pub fn read(&self, master_key: &dyn Backend) -> Result<Vec<u8>> {
+        let res = OpenOptions::new()
+            .read(true)
+            .open(self.base.join(self.name));
+        match res {
+            Ok(mut f) => {
+                let mut buf = Vec::new();
+                f.read_to_end(&mut buf)?;
+                let (_, content) = Header::parse(&buf)?;
+                let mut encrypted_content = EncryptedContent::default();
+                encrypted_content.merge_from_bytes(content)?;
+                let plaintext = master_key.decrypt(&encrypted_content)?;
+
+                Ok(plaintext)
+            }
+            Err(e) => {
+                if e.kind() == ErrorKind::NotFound {
+                    return Ok(Vec::new());
+                }
+                Err(Error::Io(e))
+            }
+        }
+    }
+
+    pub fn write(&self, plaintext_content: &[u8], master_key: &dyn Backend) -> Result<()> {
+        // Write to a tmp file.
+        // TODO what if a tmp file already exists?
+        let origin_path = self.base.join(&self.name);
+        let mut tmp_path = origin_path.clone();
+        tmp_path.set_extension(format!("{}.{}", thread_rng().next_u64(), TMP_FILE_SUFFIX));
+        let mut tmp_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(&tmp_path)
+            .unwrap();
+
+        // Encrypt the content.
+        let encrypted_content = master_key
+            .encrypt(&plaintext_content)?
+            .write_to_bytes()
+            .unwrap();
+        let header = Header::new(&encrypted_content);
```

Comment:
I'll change the header to checksum plaintext instead of ciphertext. I'll do that on my side.